src/lib.rs

```diff
@@ -0,0 +1,399 @@
+/// Raises an error if the PodSpec uses a service account that has high-risk permissions
+fn validate_pod(
+    pod_spec: apicore::PodSpec,
+    namespace: &str,
+    settings: &Settings,
+) -> Result<(), String> {
+    if pod_spec.automount_service_account_token == Some(false) {
+        debug!(
+            LOG_DRAIN,
+            "Pod does not require service account token, skipping validation."
+        );
+        return Ok(());
+    }
+
+    let service_account = pod_spec
+        .service_account_name
+        .map(|sa| format!("system:serviceaccount:{namespace}:{sa}"))
+        .unwrap_or_else(|| format!("system:serviceaccount:{namespace}:default"));
+
+    info!(
+        LOG_DRAIN,
+        "Validating pod with service account: {}", service_account
+    );
+
+    let subject_access_reviews =
+        build_subject_access_reviews(&service_account, settings.block_rules.as_slice());
+    let mut can_i_failure_counter = 0;
+
+    for sar in &subject_access_reviews {
+        let can_i_request = CanIRequest {
+            subject_access_review: sar.to_owned(),
+            disable_cache: false,
+        };
+
+        match can_i(can_i_request) {
+            Ok(status) if status.allowed => {
+                // exit immediately whe allowed to do a high risk operation,
+                // this is a performance optimization
+                return Err(build_rejection_message(
+                    &service_account,
+                    &sar.resource_attributes,
+                ));
+            }
+            Ok(_) => {
+                // nothing to do, not allowed to do high risk operations
+            }
+            Err(e) => {
+                warn!(LOG_DRAIN, "Failed to check permissions: {}", e);
+                can_i_failure_counter += 1;
+                continue;
+            }
+        }
+    }
+
+    if can_i_failure_counter == subject_access_reviews.len() {
+        // none of the SubjectAccessReviews were successful, we prefer to be cautious
+        // and reject the request
+        return Err("Failed to check permissions for all SubjectAccessReviews".to_owned());
+    }
+
+    Ok(())
+}
+
+fn build_rejection_message(
+    service_account: &str,
+    resource_attributes: &ResourceAttributes,
+) -> String {
+    let resource_operation = if let Some(namespace) = &resource_attributes.namespace {
+        format!(
+            "{} {}/{} in namespace '{}'",
+            resource_attributes.verb,
+            resource_attributes.group.clone().unwrap_or_default(),
+            resource_attributes.resource,
+            namespace,
+        )
+    } else {
+        format!(
+            "{} {}/{} in the cluster",
+            resource_attributes.verb,
+            resource_attributes.group.clone().unwrap_or_default(),
+            resource_attributes.resource
+        )
+    };
+
+    format!(
+        "Cannot use service account '{service_account}' with permissions to perform {resource_operation}",
```

Comment:
@flavio is there a reason to remove the reason from the error message? This was a review request. 

If we agree to remove it, fine. I just do not want to ignore other requests. ;)

Reply:
I think we reply the wrong thread. :smile: